client/files.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime"
+	"mime/multipart"
+	"net/url"
+)
+
+// PullOptions contains the options for a call to Pull.
+type PullOptions struct {
+	// Path indicates the absolute path of the file in the remote system
+	// (required).
+	Path string
+
+	// Target is the destination io.Writer that will receive the data (required).
+	Target io.Writer
+}
+
+type fileResult struct {
+	Path  string `json:"path"`
+	Error *Error `json:"error,omitempty"`
+}
+
+// Pull retrieves a file from the remote system.
+func (client *Client) Pull(opts *PullOptions) error {
+	query := url.Values{
+		"action": {"read"},
+		"path":   {opts.Path},
+	}
+	headers := map[string]string{
+		"Accept": "multipart/form-data",
+	}
+
+	rsp, err := client.raw(context.Background(), "GET", "/v1/files", query, headers, nil)
+	if err != nil {
+		return err
+	}
+	defer rsp.Body.Close()
+
+	// Obtain Content-Type to check for a multipart payload and parse its value
+	// in order to obtain the multipart boundary
+	contentType := rsp.Header.Get("Content-Type")
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return fmt.Errorf("invalid Content-Type: %w", err)
+	}
+	if mediaType != "multipart/form-data" {
+		// Not an error response after all
+		return fmt.Errorf("expected a multipart response but didn't get one")
+	}
+
+	mr := multipart.NewReader(rsp.Body, params["boundary"])
+	for {
```

Comment:
We're expecting one and only one file here (two multipart parts, `files` and `response`), so it seems to me we shouldn't use a `for` loop, but instead just call `NextPart()` twice in a row, ensuring we get `FormName() == "files"` for the first one, and `FormName() == "response"` for the second.

With the current approach, it will succeed silently and fall to the bottom if there are 0 or 1 parts (but there should be exactly 2).

So something like this:

```go
mr := multipart.NewReader(rsp.Body, params["boundary"])

filesPart, err := mr.NextPart()
if err != nil {
    return fmt.Errorf("cannot decode multipart payload: %w", err)
}
defer filesPart.Close()
if filesPart.FormName() != "files" {
    return fmt.Errorf(`expected first field name to be "files", got %q`, filesPart.FormName())
}
if _, err = io.Copy(opts.Target, part); err != nil {
    return fmt.Errorf("cannot write: %w", err)
}

responsePart, err := mr.NextPart()
if err != nil {
    return fmt.Errorf("cannot decode multipart payload: %w", err)
}
defer responsePart.Close()
if responsePart.FormName() != "response" {
    return fmt.Errorf(`expected second field name to be "response", got %q`, responsePart.FormName())
}
// process response metadata
```

Reply:
You are completely right about it failing silently, that must be fixed. However, my rationale for using a loop here is that I didn't want to make any assumption on the ordering of the multipart data. One could place the `response` part before `files` and it would be as valid as doing it the other way around. Your approach is more clear and surely works on Pebble, but should we impose this "hidden" restriction on future Pebble client implementations?

Reply:
Interesting rationale. However, the order is significant: the `files` part (can be multiple in the API, but in this case always 1) is guaranteed to come before the `response` part. This is a design decision, because the files have to be opened and read and written to the HTTP connection, which can error in various ways. So the errors are only known afterward trying to write the file, hence including the `response` with the metadata+errors at the bottom.

When pushing (writing), it's the opposite. The client must include the `request` metadata part first, then the `files` parts.

Reply:
Thanks for your feedback!